code/eyetracking/eyetrackingrun.py

```diff
@@ -0,0 +1,781 @@
+            print("Sampling Frequency: Not available")
+
+        row_thresholds = self.messages[
+            self.messages["trialid "].str.contains("THRESHOLDS", case=False, regex=True)
+        ].head(1)
+        thresholds_text = row_thresholds["trialid "].iloc[0]
+        print(thresholds_text, "threshold")
+        match_thresholds = re.search(r"THRESHOLDS (\w+) (\d+) (\d+)", thresholds_text)
+
+        if match_thresholds:
+            _, pupil_threshold, CR_threshold = match_thresholds.groups()
+            pupil_threshold = int(pupil_threshold)
+            CR_threshold = int(CR_threshold)
+            print("Pupil Threshold:", pupil_threshold)
+            print("CR Threshold:", CR_threshold)
+        else:
+            pupil_threshold = None
+            CR_threshold = None
+            print("Pupil Threshold: Not available")
+            print("CR Threshold: Not available")
+
+        row_fit_param = self.messages[
+            self.messages["trialid "].str.contains("ELCL_PROC", case=False, regex=True)
+        ].head(1)
+        fit_param_text = row_fit_param["trialid "].iloc[0]
+
+        if "ELLIPSE" in fit_param_text:
+            pupil_fit_method = "ellipse"
+        else:
+            pupil_fit_method = "center-of-mass"
+
+        print("Pupil Fitting Method:", pupil_fit_method)
+
+        return (
+            recorded_eye,
+            eye_tracking_method,
+            sampling_frequency,
+            pupil_threshold,
+            CR_threshold,
+            pupil_fit_method,
+        )
+
+    def extract_header(self) -> List[str]:
+        """
+        Extracts header information from the messages DataFrame.
+
+        Returns:
+        list[str]: A list of strings containing the extracted header information.
+        """
+        self.messages["trialid_cleaned"] = self.messages["trialid "].str.replace(
+            r"[\n\x00\t]", ""
+        )
+
+        record_index = self.messages[
+            self.messages["trialid_cleaned"].str.contains(
+                "RECORD", case=False, regex=True
+            )
+        ].index
+
+        if not record_index.empty:
+            header = self.messages.loc[: record_index[0], "trialid_cleaned"].tolist()
+            return header
+        return self.messages["trialid_cleaned"].tolist()
+
+    def save_and_process_samples(
+        self, BIDS_folder_path: str, include_events: bool = True
+    ) -> List[str]:
+        """
+        Save the processed samples DataFrame into a compressed tsv file and return column names.
+
+        Args:
+            BIDS_folder_path (str): Root directory path of the BIDS dataset.
+            include_events (bool, optional): Include events in the processing. Default is True.
+
+        Returns:
+            List[str]: A list of column names in the processed DataFrame.
+        """
+        if include_events:
+            self.add_events()
+
+        self.samples.loc[self.samples["pa_right"] < 1, "pa_right"] = "n/a"
```

Comment:
This is a very bad idea because it's going to cast the whole column into an object.

```suggestion
        self.samples.loc[self.samples["pa_right"] < 1, "pa_right"] = np.nan
```